quinn-proto/src/token_memory_cache.rs

```diff
@@ -0,0 +1,539 @@
+//! Storing tokens sent from servers in NEW_TOKEN frames and using them in subsequent connections
+
+use bytes::Bytes;
+use slab::Slab;
+use std::{
+    collections::{hash_map, HashMap},
+    mem::take,
+    sync::{Arc, Mutex},
+};
+use tracing::trace;
+
+use crate::token::TokenStore;
+
+/// `TokenStore` implementation that stores up to `N` tokens per server name for up to a
+/// limited number of server names, in-memory
+#[derive(Debug)]
+pub struct TokenMemoryCache<const N: usize>(Mutex<State<N>>);
```

Comment:
Does `N` really need to be a compile-time constant? Dynamic values are simpler and mean less work for the compiler downstream.

Reply:
Yeahhhhhhhhhhh I guess not. My argument of avoiding a heap allocation seems kinda flimsy since there's already going to be 1 mandatory heap allocation for the server name and then another for each token.

Reply:
Heap allocations aren't something to be avoided for their own sake, in any case.